File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/IntradayTickElementDataArray.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/IntradayTickElementDataArray.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Datetime;
import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.Map.Entry;

public class IntradayTickElementDataArray extends Element
{
	private List<IntradayTickElementData> _tickData;
	
	IntradayTickElementDataArray(Map<Datetime, Tuple3<String, Double, Integer>> ticks, boolean includeConditionCodes) throws Exception
	{
		this._tickData = new ArrayList<IntradayTickElementData>();
		
		for(Entry<Datetime, Tuple3<String, Double, Integer>> h : ticks.entrySet())
		{
			Datetime dt = h.getKey();
			Tuple3<String, Double, Integer> tpl = h.getValue();
			String str = tpl.item1();
			Double dd = tpl.item2();
			Integer it = tpl.item3();
			
			IntradayTickElementData elmFieldData = new IntradayTickElementData(dt, str, dd, it, includeConditionCodes);
			this._tickData.add(elmFieldData);
		}
	}
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("IntradayTickData"));
    }
	
	public Element getValueAsElement(int index)
	{
		return this._tickData.get(index);
	}

	public Name name() throws Exception
	{
		return new Name("tickData");
	}

	public int numValues()
	{
		return this._tickData.size();
	}

	public int numElements()
	{
		return 0;
	}
	
    public boolean isComplexType()
    {
    	return false;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    public Object getValue(int index)
    {
    	return this._tickData.get(index);
    }
	
	protected StringBuilder prettyPrint(int tabIndent) throws Exception
	{
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

		result.append(String.format("%s%s[] = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));        
        for(int i = 0; i < this._tickData.size(); i++)
        {
        	result.append(this._tickData.get(i).prettyPrint(tabIndent + 1));
        }
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        return result;
	}
    
    
}

File: BEmu/IntradayTickRequest/IntradayTickElementDataArray.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="IntradayTickRequest/IntradayTickElementDataArray.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.IntradayTickRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class IntradayTickElementDataArray : Element
    {
        private readonly List<IntradayTickElementData> _tickData;

        internal IntradayTickElementDataArray(SortedDictionary<DateTime, Types.Tuple3<string, double, int>> ticks, bool includeConditionCodes)
        {
            this._tickData = new List<IntradayTickElementData>();

            foreach (var dtDate in ticks)
	        {
                IntradayTickElementData elmFieldData = new IntradayTickElementData(dtDate.Key, dtDate.Value, includeConditionCodes);
                this._tickData.Add(elmFieldData);
	        }
        }

        public override Element GetValueAsElement(int index) { return (Element)this._tickData[index]; }
        public override object this[int index] { get { return this._tickData[index]; } }

        public override SchemaTypeDefinition TypeDefinition { get { return new SchemaTypeDefinition(this.Datatype, new Name("IntradayTickData")); } }
        public override Name Name { get { return new Name("tickData"); } }
        public override int NumElements { get { return 0; } }
        public override int NumValues { get { return this._tickData.Count; } }
        public override bool IsComplexType { get { return false; } }
        public override bool IsArray { get { return true; } }
        public override bool IsNull { get { return false; } }
        public override IEnumerable<Element> Elements { get { if (this._tickData.Count < -1) yield return null; } } //This is the only way I can think to have the compiler output "Enumeration yielded no results"

        public override object GetValue(int index)
        {
            return this._tickData[index];
        }

        internal override StringBuilder PrettyPrint(int tabIndent, bool surroundValueWithQuotes = false)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}{1}[] = {{{2}", tabs, this.Name, Environment.NewLine);
            foreach (var item in this._tickData)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);

            return result;
        }
    }
}
